### PR TITLE
Remote: Fix "file not found" error when remote cache is changed from enabled to disabled.

### PR DESCRIPTION
**Pull request by coeuvre**
When use BwtB, intermediate outputs are not downloaded. If a following build disables remote cache, a "file not found" error will be thrown if an action doesn't have its inputs downloaded in previous build because those files cannot be downloaded in this build.

This change fix this issue by:
  1. Do not load remote metadata from action cache if `--experimental_action_cache_store_output_metadata` is set and remote cache is disabled.
  2. Invalidate action nodes if previous build use BwtB and remote cache is changed from enabled to disabled.

Fixes https://github.com/bazelbuild/bazel/issues/13882.

### CODE REVIEW

File: src/test/shell/bazel/remote/remote_execution_test.sh [hunk 1]
```diff
@@ -3204,4 +3204,72 @@ EOF
   expect_not_log "WARNING: Writing to Remote Cache:"
 }
 
+function test_download_toplevel_when_turn_remote_cache_off() {
+  # Test that BwtB does cause build failure if remote cache is disabled in a following build.
+  # See https://github.com/bazelbuild/bazel/issues/13882.
+
+  mkdir -p a
+  cat > a/BUILD <<EOF
```
Comment (alexjski) [inline src/test/shell/bazel/remote/remote_execution_test.sh:3212]: nit: you can use `EOF` and not need to escape the `$` in the content.

File: src/test/shell/bazel/remote/remote_execution_test.sh [hunk 2]
```diff
@@ -3204,4 +3204,72 @@ EOF
   expect_not_log "WARNING: Writing to Remote Cache:"
 }
 
+function test_download_toplevel_when_turn_remote_cache_off() {
+  # Test that BwtB does cause build failure if remote cache is disabled in a following build.
+  # See https://github.com/bazelbuild/bazel/issues/13882.
+
+  mkdir -p a
```
Comment (alexjski) [inline src/test/shell/bazel/remote/remote_execution_test.sh:3211]: supernit: `-p` not needed in this case

File: src/test/shell/bazel/remote/remote_execution_test.sh [hunk 3]
```diff
@@ -3204,4 +3204,72 @@ EOF
...
+  echo 'foo' > a/in.txt
+
+  # populate the cache
+  bazel build \
+    --remote_cache=grpc://localhost:${worker_port} \
+    --remote_download_toplevel \
+    --verbose_failures \
+    //a:consumer >& $TEST_log || fail "Failed to populate the cache"
+
+  bazel clean || fail "Failed to clean"
```
Comment (alexjski) [inline src/test/shell/bazel/remote/remote_execution_test.sh:3235]: super nit: `bazel clean >& "${TEST_log}" || fail ...`

File: src/test/shell/bazel/remote/remote_execution_test.sh [hunk 4]
```diff
@@ -3204,4 +3204,72 @@ EOF
...
+
+  # download top level outputs without remote metadata
+  bazel build \
+    --remote_cache=grpc://localhost:${worker_port} \
+    --remote_download_toplevel \
+    --verbose_failures \
+    //a:consumer >& $TEST_log || fail "Failed to download outputs without remote metadata"
+  (! [[ -f bazel-bin/a/a.txt ]] && ! [[ -f bazel-bin/a/b.txt ]]) \
```
Comment (alexjski) [inline src/test/shell/bazel/remote/remote_execution_test.sh:3243]: nit: can we reverse that to `[[ -f bazel-bin/a/a.txt ]] || [[ -f bazel-bin/a/b.txt ]] && fail ...`?

File: src/test/shell/bazel/remote/remote_execution_test.sh [hunk 5]
```diff
@@ -3204,4 +3204,72 @@ EOF
...
+  || fail "Expected outputs of producer are not downloaded without remote metadata"
+
+  # build without remote cache without remote metadata
```
Comment (alexjski) [inline src/test/shell/bazel/remote/remote_execution_test.sh:3246]: Is this the build which fails without your change? The reason I am asking is because this test may be merging 2 test cases in here. In general, it is a little hard to follow what is arrange-act-assert in here.

Meta-comment -- is there a way to run the remote tests in a Java test using BuildIntegrationTestCase? In those, it is much easier to parameterize etc.

**Review by alexjski — CHANGES_REQUESTED** [review]
I would recommend splitting that into 2 logically separate changes as in the description.

File: src/test/shell/bazel/remote/remote_execution_test.sh (hunk 2 above)
Comment (coeuvre) [reply src/test/shell/bazel/remote/remote_execution_test.sh:3211]: Done.

File: src/test/shell/bazel/remote/remote_execution_test.sh (hunk 1 above)
Comment (coeuvre) [reply src/test/shell/bazel/remote/remote_execution_test.sh:3212]: Done.

File: src/test/shell/bazel/remote/remote_execution_test.sh (hunk 3 above)
Comment (coeuvre) [reply src/test/shell/bazel/remote/remote_execution_test.sh:3235]: Done.

File: src/test/shell/bazel/remote/remote_execution_test.sh (hunk 4 above)
Comment (coeuvre) [reply src/test/shell/bazel/remote/remote_execution_test.sh:3243]: Done.

File: src/test/shell/bazel/remote/remote_execution_test.sh (hunk 5 above)
Comment (coeuvre) [reply src/test/shell/bazel/remote/remote_execution_test.sh:3246]: Split into 2 test cases.

> Meta-comment -- is there a way to run the remote tests in a Java test using BuildIntegrationTestCase? In those, it is much easier to parameterize etc.

Not sure. These remote tests run a remote worker process which is a real remote server. Do we prefer BuildIntegrationTestCase over shell in general? If so, I can try to make it work with remote tests and write new tests there in future PRs.

**Review by coeuvre — COMMENTED** [review]
Thanks for the comments. I will split this PR when importing.

File: src/test/shell/bazel/remote/remote_execution_test.sh [hunk 6]
```diff
@@ -3232,44 +3232,73 @@ EOF
...
   echo 'bar' > a/in.txt
   bazel build \
     --remote_download_toplevel \
     --verbose_failures \
-    //a:consumer >& $TEST_log || fail "Failed to build without remote cache without remote metadata"
+    //a:consumer >& $TEST_log || fail "Failed to build without remote cache"
+}
 
-  bazel clean || fail "Failed to clean"
+function test_download_toplevel_when_turn_remote_cache_off_with_metadata() {
```
Comment (alexjski) [inline src/test/shell/bazel/remote/remote_execution_test.sh:3254]: These 2 test cases differ by 1 option only -- why not put the test case as a separate helper function and call that twice with different parameters (Bash tests parameterization). Example [here](https://github.com/bazelbuild/bazel/blob/bb2e0ae6021da27cb4d6044bcb5c26e00d2b3db3/src/test/shell/integration/client_sigint_test.sh#L80-L86).

File: src/test/shell/bazel/remote/remote_execution_test.sh [hunk 7]
```diff
@@ -3232,44 +3232,73 @@ EOF
...
 
-  bazel clean || fail "Failed to clean"
+  bazel clean >& $TEST_log || fail "Failed to clean"
 
-  # download top level outputs without remote metadata
+  # download top level outputs
   bazel build \
     --remote_cache=grpc://localhost:${worker_port} \
     --remote_download_toplevel \
     --verbose_failures \
```
Comment (alexjski) [inline src/test/shell/bazel/remote/remote_execution_test.sh:3241]: supernit: `--verbose_failures` is unlikely needed/we could add that to the blazerc for the whole test if that aids debugging failures.

File: src/test/shell/bazel/remote/remote_execution_test.sh (hunk 5 above)
Comment (alexjski) [reply src/test/shell/bazel/remote/remote_execution_test.sh]: We generally do. I wouldn't tell you to rewrite these tests, but you may want to prefer Java-based ones when applicable in the future. Sent you a document talking about that offline.

**Review by alexjski — APPROVED** [review]

File: src/test/shell/bazel/remote/remote_execution_test.sh (hunk 6 above)
Comment (coeuvre) [reply src/test/shell/bazel/remote/remote_execution_test.sh:3211]: Done.

File: src/test/shell/bazel/remote/remote_execution_test.sh (hunk 7 above)
Comment (coeuvre) [reply src/test/shell/bazel/remote/remote_execution_test.sh]: Done.

File: src/test/shell/bazel/remote/remote_execution_test.sh [hunk 8]
```diff
@@ -3286,8 +3250,7 @@ EOF
   bazel build \
     --remote_cache=grpc://localhost:${worker_port} \
     --remote_download_toplevel \
-    --experimental_action_cache_store_output_metadata \
-    --verbose_failures \
+    $extra_build_flags \
```
Comment (alexjski) [inline src/test/shell/bazel/remote/remote_execution_test.sh:3253]: I suspect bash linter may not like that -- your best bet may be to use the "$@" array directly or declare and use extra_build_flags as an array and use like that (personally, I recommend `"$@"`):

```
local -r extra_build_flags=("$@")
...
"${extra_build_flags[@]}"
```

**Review by alexjski — APPROVED** [review]

**Review by alexjski — APPROVED** [review]
